### ISSUE TITLE
Embroider don't work with Ember Engines

### ISSUE THREAD

**Issue by villander** [embroider-build/embroider#320]
A reproduction - https://github.com/villander/ember-monorepo-demo/tree/embroider

**Comment by stefanpenner** [comment]
@villander could you also provide stack trace / description of the issue?

**Comment by villander** [comment]
@stefanpenner 

```
Uncaught Error: Failed to load asset manifest. For browser environments, verify the meta tag with name "host-app/config/asset-manifest" is present. For non-browser environments, verify that you included the node-asset-manifest module.
    at eval (asset-manifest.js:22)
    at Module../config/asset-manifest.js (chunk.65718aac7904ab65649b.js:393)
    at __webpack_require__ (chunk.76ec168998ef1526eb69.js:79)
    at eval (load-asset-manifest.js:3)
    at Module../instance-initializers/load-asset-manifest.js (chunk.65718aac7904ab65649b.js:465)
    at __webpack_require__ (chunk.76ec168998ef1526eb69.js:79)
    at Module.eval [as callback] (host-app.js:19)
    at Module.exports (loader.js:106)
    at requireModule (loader.js:27)
    at resolveInitializer (index.js:8)
```

**Comment by ef4** [comment]
Does this still happen on embroider master? I think this may be the same as
an already fixed issue.

**Comment by stefanpenner** [comment]
@ef4 I'll take a stab at adding engines to macro-app, so we can get signals early. I am also running into engine related issues.

**Comment by villander** [comment]
@stefanpenner we need as well figure out how embroider will work with engines, maybe some documentation for this? I can help. We just will make lazy loading by routes/components into engines? can we replace `ember-asset-loader` with `embroider`?

**Comment by stefanpenner** [comment]
@villander ya, Im not 100% sure either. @ef4 can hopefully provide us some context.

**Comment by villander** [comment]
> Does this still happen on embroider master? I think this may be the same as
an already fixed issue.

yup @ef4

**Comment by villander** [comment]
> we need as well figure out how embroider will work with engines, maybe some documentation for this? I can help. We just will make lazy loading by routes/components into engines? can we replace ember-asset-loader with embroider?

@ef4 any direction about it?

**Comment by ef4** [comment]
> we need as well figure out how embroider will work with engines, maybe some documentation for this?

I don't think this is different from any of the other hundred compatibility things we've had to implement. It doesn't rise to the level of needing documentation. There won't be anything app authors are asked to change.

Engines aren't very different from any other addons. They have a small number of extra behaviors. One of those behaviors is laziness, but embroider already contains examples of how to deal with laziness (see the route-based code splitting).

I can certainly teach more about this area, but I need to balance that time investment. The more you can spend time first on your own debugging and understanding how things work, the more effectively I can help guide. For example, this particular issue is related to embroider not running the postprocessTree addon hook. Dig in until you can see what I mean by that, and then come up with a hack to make the crash go away. From there we can have a more productive discussion about next steps.

**Comment by villander** [comment]
> I don't think this is different from any of the other hundred compatibility things we've had to implement. It doesn't rise to the level of needing documentation. There won't be anything app authors are asked to change.

I should have been more explicit, sorry about it. My questions is how something will work with engines, like:

`staticComponents`, `packageRules` in embroider codebase and some docs meaning (app and all addons) at build time. This already include engines standalone and in-repo? 

> Engines aren't very different from any other addons. They have a small number of extra behaviors. One of those behaviors is laziness, but embroider already contains examples of how to deal with laziness (see the route-based code splitting).

I know how embroider deal with laziness, the code is very similar with ember-engines code. My question abou some documentation is about `ember-engines/routes.js` for example a different file to map routes inside engines. How can I split routes inside a engine standalone (we don't have `ember-cli-build.js` here)? 

> I can certainly teach more about this area, but I need to balance that time investment. The more you can spend time first on your own debugging and understanding how things work, the more effectively I can help guide. For example, this particular issue is related to embroider not running the postprocessTree addon hook. Dig in until you can see what I mean by that, and then come up with a hack to make the crash go away. From there we can have a more productive discussion about next steps.

Ok, I will dig a lot to advance, thanks

**Comment by ef4** [comment]
> My questions is how something will work with engines, like: staticComponents, packageRules

Ah yes, those are good questions. 

`packageRules` are already scoped to packages, which engines are, so nothing changes there.

Because of the isolation in engines, it would probably be possible to make `staticComponents` a per-engine setting. I'll look into that as I implement the rest of engine support.
